geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeatureStore.scala

```diff
@@ -0,0 +1,177 @@
+
+  private lazy val contentState = entry.getState(getTransaction).asInstanceOf[CassandraContentState]
+
+  override def getWriterInternal(query: Query, flags: Int): FW[SimpleFeatureType, SimpleFeature] = {
+    if((flags | WRITER_ADD) == WRITER_ADD) new AppendFW(contentState.sft, contentState.session)
+    else                                   new UpdateFW(contentState.sft, contentState.session)
+  }
+
+  override def buildFeatureType(): SimpleFeatureType = contentState.sft
+
+  override def getBoundsInternal(query: Query): ReferencedEnvelope = new ReferencedEnvelope(-180.0, 180.0, -90.0, 90.0, DefaultGeographicCRS.WGS84)
+
+  override def getCountInternal(query: Query): Int = {
+    // TODO: might overflow
+    if(query.equals(Query.ALL) || FilterHelper.isFilterWholeWorld(query.getFilter)) {
+      contentState.session.execute(contentState.ALL_COUNT_QUERY.bind()).iterator().next().getLong(0).toInt
+    } else {
+      val plans = planQuery(query)
+      executeGeoTimeCountQuery(query, plans).toInt
+    }
+  }
+
+  case class RowAndColumnQueryPlan(row: Int, lz3: Long, uz3: Long, contained: Boolean)
+
+  def executeGeoTimeCountQuery(query: Query, plans: Seq[RowAndColumnQueryPlan]) = {
+    // TODO: currently overestimates the count in order to increase performance
+    // Fix overestimation by pushing geo etc predicates down into the database
+    val features = contentState.builderPool.withResource { builder =>
+      val futures = plans.map { case RowAndColumnQueryPlan(r, l, u, contained) =>
+        val q = contentState.GEO_TIME_COUNT_QUERY.bind(r: Integer, l: lang.Long, u: lang.Long)
+        contentState.session.executeAsync(q)
+      }
+      futures.flatMap { f => f.get().iterator().toList }.map(_.getLong(0)).sum
+    }
+    features
+  }
+
+
+  override def getReaderInternal(query: Query): FeatureReader[SimpleFeatureType, SimpleFeature] = {
+    val iter =
+      if(query.equals(Query.ALL) || FilterHelper.isFilterWholeWorld(query.getFilter)) {
+        getAllFeatures
+      } else {
+        val plans    = planQuery(query)
+        val features = executeGeoTimeQuery(query, plans)
+        features.iterator
+      }
+    new DelegateSimpleFeatureReader(contentState.sft, new DelegateSimpleFeatureIterator(iter))
+  }
+
+  def executeGeoTimeQuery(query: Query, plans: Seq[RowAndColumnQueryPlan]): Seq[SimpleFeature] = {
+    val features = contentState.builderPool.withResource { builder =>
+      val futures = plans.map { case RowAndColumnQueryPlan(r, l, u, contained) =>
+        val q = contentState.GEO_TIME_QUERY.bind(r: Integer, l: lang.Long, u: lang.Long)
+        (contained, contentState.session.executeAsync(q))
+      }
+      futures.flatMap { case (contains, fut) =>
+        postProcessResults(query, builder, contains, fut)
+      }
+    }
+    features
+  }
+
+  val WHOLE_WORLD = new ReferencedEnvelope(-180.0, 180.0, -90.0, 90.0, DefaultGeographicCRS.WGS84)
```

Comment:
Yeah, i've looked for this in geotools somewhere.  It's gotta be in there but I always end up giving up the search and just building my own.